### ISSUE TITLE
Is MariaDB supported? 

### ISSUE THREAD

**Issue by johnkarasev** [diesel-rs/diesel#1882]
Just wondering if diesel supports MariaDB.

**Comment by weiznich** [comment]
`MariaDB` is more or less supported through the `MySql` backend. 

In the long term we want to add a distinct `MariaDB` backend that  shares most of it implementation with the existing `MySql` backend. 

If anyone is willing working on this feel free to claim that issue or drop a message on gitter.

**Comment by ian-p-cooke** [comment]
@virome are you still working on this?  I might want to do something similar to support TiDB and would like to see how you refactor things.

**Comment by virome** [comment]
@ian-p-cooke Unfortunately, I ended up not having much time to work on this. If no one else wants to take it, I still would like to finish it at some point.
If you're ready to go with your work, I wouldn't wait for me.

**Comment by p-alik** [comment]
I used a workaround to build diesel with mariadb
https://stackoverflow.com/questions/61505861/crate-with-diesel-and-mysql-dependencies-no-longer-compiles-against-mariadb-afte

**Comment by saresend** [comment]
Is this still available? I'd be interested to give it a go

**Comment by weiznich** [comment]
@saresend This is still an open issue, so feel free to work on this. Be warned that this requires likely a complex large change to diesel itself.

**Comment by saresend** [comment]
Yeah that's fine, just want to double check you'd be willing to help answer questions / support this work

**Comment by weiznich** [comment]
@tgross35 Let's move the discussion about a custom mariadb backend here:

> Thanks for the response, seems like there's a couple reasons for a dedicated mariadb backend. What would be the bare minimum PR to get support moving in the right direction? It seems like duplicating diesel/diesel/src/mysql and updating the connection would be needed, but I'm not sure how many of the other sections of the repo would need to be touched just to get something going, or what a prioritized todo list would look like.

It's hard to tell what's the exact minimum PR for something like that, because as already written: We do not even have a specific design for that in mind. Keep that in mind for anything else that follows. I would prefer something that does not duplicate that much code, as that all is code that needs to be maintained. I would guess that there is no need to duplicate all of `diesel/diesel/src/mysql`. For example the connection/serialization/deserialization implementation can and should be shared between a potential mariadb backend and the existing mysql backend. I would likely start with the following approach:

* Add a new mariadb type and implement `Backend` (and related traits) for this. Essentially that should duplicate [this file](https://github.com/diesel-rs/diesel/blob/9fea8dfb98389acde9bd5c1666131da1c8c8ce24/diesel/src/mysql/backend.rs). You can likely just copy that as a starting point, although we might want to tweak the `SqlDialect` impl at some point.
* Add support for the new backend in the [`SqlType` derive](https://github.com/diesel-rs/diesel/blob/dba995346b9b38984ea6f1129c2ea7193abfa0bc/diesel_derives/src/sql_type.rs). Again that could copy most of the existing mysql implementation there.
* Go through all the places where `#[diesel(mysql_type = …)]` is used and add an analogus `#[diesel(mariadb_type = …)]`
* Make the `MysqlConnection` type generic over the backend and ensure that it only accepts `Mysql` and `MariaDB` as potential backends.
* Check all [mysql specific queryfragment impls](https://github.com/search?q=repo%3Adiesel-rs%2Fdiesel%20QueryFragment%3CMysql&type=code) and implement variants for mariadb
* Check [all the existing (de)serialization impls](https://github.com/diesel-rs/diesel/tree/9fea8dfb98389acde9bd5c1666131da1c8c8ce24/diesel/src/mysql/types) and make them applicable for both the mariadb and mysql backend.

All of that is quite a bit of work, so don't expect any fast changes there. Additionally we would need some testing strategy. We likely cannot afford testing another backend with the current setup.

**Comment by Mrodent** [comment]
Posted a Stack Overflow question about Rust and MariaDB today. Anyone interested in developing this "backend for MariaDB" at the current time?

**Comment by weiznich** [comment]
@Mrodent As our issue tracker is a place to track issues and not a place to provide support I will answer only the part of your question that is actually related to the development of diesel. For anything else (especially anything that is related to your "setup problems") please use our support channels (or just refer to the answers already given in other places.)

> Anyone interested in developing this "backend for MariaDB" at the current time?

We are still open to accept contributions here. Otherwise this is nothing that I plan to spend time on implementing it on my own in the next time as I personally do neither use mysql (nor mariadb). So this is something that needs to be implemented by an interested contributor. 

> Isn't MariaDB big enough now to deserve a worthy coverage... ?

It's not about the size of the database community, but rather about if people that use these database are interested in putting in the work to implement it. Please don't expect me to put in that work for something I won't use instead of fixing bugs in things I actually use or implement new features I actually need.
I believe that in the end it boils down to: The existing support for MariaDB via the existing MySQL backend is good enough for most of our users that nobody is willing to put in the required work to have a distinct backend. (I would estimate it with roughly several weeks of effort to implement that + getting it merged)

**Comment by Mrodent** [comment]
@weiznich Thanks for your explanations about the way to proceed with this. I'm using the ODBC workaround for the moment. But when I have a bit more time I'll try to see how far I can get with Diesel + MariaDB + W10. If I have any success I'll report back with any findings, for anyone similarly facing this situation.

**Comment by ottok** [comment]
FYI: Related https://github.com/sgrif/mysqlclient-sys/issues/47 by @robertsilen was closed via https://github.com/sgrif/mysqlclient-sys/pull/49

**Comment by weiznich** [comment]
@ottok We are happy to accept a PR that makes it more clear that you can also use libmariadb with the mysqlclient-sys. 
Otherwise this is not really related to what's described in this issue as that's only a client side thing while this issue talks about supporting sql syntax send to the server.

**Comment by KelvinxG** [comment]
has anyone working on this? I might propose some minimal user interfaces MariaDB for Rust users.

**Comment by weiznich** [comment]
@KelvinxG We are still looking for contributions here

**Comment by ottok** [comment]
This could be a chicken-and-egg problem? Devs using MariaDB might opt to use other libraries as the README of diesel states:

> Supported databases:
> 
> 1. [PostgreSQL](https://docs.diesel.rs/master/diesel/pg/index.html)
> 2. [MySQL](https://docs.diesel.rs/master/diesel/mysql/index.html)
> 3. [SQLite](https://docs.diesel.rs/master/diesel/sqlite/index.html)

If this mentioned MariaDB is supported via the MySQL backend, and that contributions for a distinct backend are welcome in #1882, a contributor might surface..? Just my 2 cents.

**Comment by weiznich** [comment]
@ottok Mentioning MariaDB there is not correct in my opinion as we really do not support any MariaDB specific features. For PostgreSQL we also don't mention any postgres wire compatible database that happens to work. This list really should only mention databases that are tested and guaranteed to work. As written several times before: We are open to consider accepting a backend for this, but until that happens the list stays as it is written. Bringing up the topic again and again will not change anything, especially if it's raised from persons closely related to MariaDB itself. Consider this as a warning this time that this kind of marketing comments are not welcome here.

**Comment by weiznich** [comment]
Fixed by #5139
